### ISSUE TITLE
npm install doesn't respect package.json for custom modules. causes re-build even if package is not in package.json

### ISSUE THREAD

**Issue by Marak** [npm/npm#9762]
`npm 3.3.3`
`node v0.12.7` ( and probably all versions )

Not sure if this is intended behavior or not. It's unexpected.

**Scenario:**

User has run `npm install spellcheck` inside a project, but has not used `--save` parameter or included `spellcheck` module in `package.json`

`spellcheck` is now installed in the system. `node-gyp rebuild` was called.

Now, the user runs `npm install`. Despite `spellcheck` not being present in `package.json`, it is re-built. `node-gyp rebuild` is called again.

**Comment by Marak** [comment]
Is there any response to this?

This can cause a real loss of time having to wait for unnecessary re-builds..

**Comment by iarna** [comment]
Hey, you'll be happy to know this was fixed in 3.3.7, back on 2015-10-08! I'm gonna close this, but if you DO see this in more recent npm's, please open a new issue and we'll fix that too.

**Comment by Marak** [comment]
Where is the associated commit? How exactly was this fixed? What the cause of the problem?

Why is it that I had to complain about npm the company before someone even looked at this?

**Comment by ashleygwilliams** [comment]
@Marak if you look at the [`CHANGELOG.md` entry for `3.3.7`](https://github.com/npm/npm/blob/master/CHANGELOG.md#v337-2015-10-08) the problem and the solution and the links to the commits are under the second heading ["OPTIONAL DEPS, MORE OPTIONAL"](https://github.com/npm/npm/blob/master/CHANGELOG.md#optional-deps-more-optional).

**Comment by Marak** [comment]
Thank you for the information @ashleygwilliams

Still would have been nice to hear about this last year, instead of downgrading to `npm2` and researching alternatives like `ied`.

I guess this is what happens when a package manager gets millions of dollars of funding. Awesome.

**Comment by iarna** [comment]
Q1: See our release notes. https://github.com/npm/npm/releases/tag/v3.3.7 It's explained _in detail_.

Q2: You mentioned having "given up" on getting hook.io problems fixed, so I was trying to figure out what I'd missed, as I wasn't aware of any. This was the only actionable thing of yours I found, and it was fixed, so I wasn't gonna leave it open, obvs. Unfortunately, when we fix issues we don't always find ALL of the reports of them, so we don't get to close them at the time they happen. This is largely a side effect of having the volume of issues coming in that we do (mostly support requests), and while this is a high priority for our team, getting things into a better place is not something that will happen over night.
